using_restricted_array.rb

```diff
@@ -6,19 +6,58 @@
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
+
+=begin
+4. search - which looks for a given integer value in the array. Returns true if found, false otherwise. Note: do not assume that the array is sorted.
+
+5. binary_search - which looks for a given integer value in the array. Returns true if found, false otherwise. Note: Assume that the array is sorted in ascending order.
+
+6. find_largest - Finds and returns the largest value element in the integer array.
+
+7. find_smallest - Finds and returns the smallest value element in the integer array.
+=end
+
 def length(array)
-  puts "NOT IMPLEMENTED"
+  index = 0
+
+  while true
+    if array[index] != nil
+      index += 1
+    else
+      return index
+    end
+  end
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  index = 0
+  array_printed_string = ""
+
+  while true
+    if array[index] != nil
+      array_printed_string = array_printed_string.concat(array[index].to_s + " ")
```

Comment:
If instead of creating array_printed_string, you did a print as you went through the loop, the space complexity would have been O(1). Because you are creating array_printed_string and the size of it depends on the size of the input array, the space complexity for this method, as it currently stands is O(n).